app/src/main/java/org/eyeseetea/malariacare/database/iomodules/dhis/importer/PullController.java

```diff
@@ -0,0 +1,311 @@
+        for (OrganisationUnit assignedOrganisationsUnit : assignedOrganisationsUnits) {
+            if (!ProgressActivity.PULL_IS_ACTIVE) return;
+            OrganisationUnitExtended organisationUnitExtended = new OrganisationUnitExtended(assignedOrganisationsUnit);
+            organisationUnitExtended.accept(converter);
+        }
+
+    }
+
+    /**
+     * Turns events and datavalues into
+     *
+     * @param converter
+     */
+    private void convertDataValues(ConvertFromSDKVisitor converter) {
+        if (!ProgressActivity.PULL_IS_ACTIVE) return;
+        Program appProgram = Program.getFirstProgram();
+        postProgress(context.getString(R.string.progress_pull_surveys));
+        //XXX This is the right place to apply additional filters to data conversion (only predefined orgunit for instance)
+        //For each unit
+        for (OrganisationUnit organisationUnit : MetaDataController.getAssignedOrganisationUnits()) {
+            //Each assigned program
+            for (org.hisp.dhis.android.sdk.persistence.models.Program program : MetaDataController.getProgramsForOrganisationUnit(organisationUnit.getId(), ProgramType.WITHOUT_REGISTRATION)) {
+
+                //Only events for the right program are loaded
+                if(!appProgram.getUid().equals(program.getUid())){
+                    continue;
+                }
+
+                List<Event> events = TrackerController.getEvents(organisationUnit.getId(), program.getUid());
+                Log.i(TAG, String.format("Converting surveys and values for orgUnit: %s | program: %s", organisationUnit.getLabel(), program.getDisplayName()));
+                for (Event event : events) {
+                    if (!ProgressActivity.PULL_IS_ACTIVE) return;
+                    EventExtended eventExtended = new EventExtended(event);
+
+                    //Only last 3 months
+                    if(eventExtended.isTooOld()) return;
+                    eventExtended.accept(converter);
+                }
+            }
+        }
+
+    }
+
+    /**
+     * Notifies a progress into the bus (the caller activity will be listening)
+     *
+     * @param msg
+     */
+    private void postProgress(String msg) {
+        Dhis2Application.getEventBus().post(new SyncProgressStatus(msg));
+    }
+
+    /**
+     * Notifies an exception while pulling
+     *
+     * @param ex
+     */
+    private void postException(Exception ex) {
+        Dhis2Application.getEventBus().post(new SyncProgressStatus(ex));
+    }
+
+    /**
+     * Notifies that the pull is over
+     */
+    private void postFinish() {
+        //Fixme maybe it is not the best place to reload the logged user.(Without reload the user after pull, the user had diferent id and application crash).
+        User user = User.getLoggedUser();
+        Session.setUser(user);
+        try {
+            Dhis2Application.getEventBus().post(new SyncProgressStatus());
+        } catch (Exception e) {
+            e.printStackTrace();
```

Comment:
@adrianq, I would remove the try/catch.

If you get this far the pull process is completed, this is just a mechanism to communicate (via bus) the progress activity to move on and show a proper dialog. An exception at this point should never happen.

If this happens there is something bigger not related to the pull itself and the app will break sooner or later.